packages/react/polymorphic/src/forwardRefWithAs.ts

```diff
@@ -33,31 +35,26 @@ interface ForwardRefComponent<
     MergeProps<IntrinsicElementString, OwnProps & { as?: IntrinsicElementString }>
   > {
   /**
-   * When passing an `as` prop as a component, use this overload.
-   * Merges orignal own props (without DOM props) and the inferred props
+   * When passing an `as` prop as a string, use this overload.
+   * Merges original own props (without DOM props) and the inferred props
    * from `as` element with the own props taking precendence.
    *
-   * We don't use `React.ComponentType` here as we get type errors
-   * when consumers try to do inline `as` components.
+   * We explicitly define a `JSX.IntrinsicElements` overload so that
+   * events are typed for consumers.
    */
-  <
-    As extends React.ElementType<any>,
-    // Inferring with props so inline `as` components get implicit `any` errors
-    // e.g. `as={(props) => {}}` <-- `props` errors
-    _AsWithProps = As extends React.ElementType<infer P> ? React.ElementType<P> : never
-  >(
-    props: MergeProps<_AsWithProps, OwnProps & { as: _AsWithProps }>
+  <As extends keyof JSX.IntrinsicElements = NarrowIntrinsic<IntrinsicElementString>>(
+    props: MergeProps<As, OwnProps & { as: As }>
   ): React.ReactElement | null;
 
   /**
-   * When passing an `as` prop as a string, use this overload.
-   * Merges orignal own props (without DOM props) and the inferred props
+   * When passing an `as` prop as a component, use this overload.
+   * Merges original own props (without DOM props) and the inferred props
    * from `as` element with the own props taking precendence.
    *
-   * We explicitly define a `JSX.IntrinsicElements` overload so that
-   * events are typed for consumers.
+   * We don't use `React.ComponentType` here as we get type errors
+   * when consumers try to do inline `as` components.
```

Comment:
Is the bit about inline `as` still relevant? Considering we've given up on the implicit any on this case?